core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -131,56 +153,14 @@ class TiBatchWrite(@transient val df: DataFrame,
     // TODO: lock table
     // pending: https://internal.pingcap.net/jira/browse/TIDB-1628
 
-    // TODO: if this write is update, TiDB reuses row_id. We need adopt this behavior.
-    // TODO: 1. fill column with its default value if it is null
-    //       2. refactor the writing process
-
-    // when primary is handle, it does not require allocate ids for each row.
-    val offset = if (!tiTableInfo.isPkHandle) {
-      val rowIDAllocator = RowIDAllocator.create(
-        tiDBInfo.getId,
-        tiTableInfo.getId,
-        tiSession.getCatalog,
-        tiTableInfo.isAutoIncColUnsigned,
-        df.count
-      )
-      rowIDAllocator.getStart
-    } else {
-      0
-    }
-
-    // i + start will be handle id if primary key is not handle
-    val tiKVRowRDD = df.rdd.zipWithIndex.map {
-      case (row, i) =>
-        sparkRow2TiKVRow(i + offset, row)
-    }
-
-    // deduplicate
-    val deduplicateRDD = deduplicate(tiKVRowRDD)
-
-    // encode TiROW
-    val encodedTiRowRDD = deduplicateRDD.map {
-      case (key, tiRow) => (key, tiRow, encodeTiRow(tiRow))
-    }
+    val tiRowRdd = df.rdd.map(row => sparkRow2TiKVRow(row))
 
-    // region pre-split
-    if (enableRegionPreSplit) {
-      logger.info("region pre split is enabled.")
-      val sampleRDD =
-        encodedTiRowRDD.sample(withReplacement = false, fraction = options.sampleFraction)
-      val (dataSize, sampleRDDCount) = estimateDataSize(sampleRDD, options)
+    val deduplicatedTiRowRdd = deduplicateIfNecessary(tiRowRdd, tiTableInfo.isPkHandle)
```

Comment:
what about `just throw error if primary key conflicts`?

Reply:
I am kinda confused. That is what do we in this PR.

Reply:
i mean:
1. keep check primary key duplicate logic
2. if find primary key duplicate, just throw error instead of deduplicate

because we do not deduplicate for non-primary unique key index

Reply:
For now, region pre-split only support pk is handle case.